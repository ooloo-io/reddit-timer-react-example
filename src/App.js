import React from 'react';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import GlobalStyle from './GlobalStyle';
import theme from './theme';

function App() {
  return (
    <ThemeProvider theme={theme}>
      <Router>
        <GlobalStyle />
        <Switch>
          <Route path="/search">Search</Route>
          <Route path="/">Home</Route>
        </Switch>
      </Router>
    </ThemeProvider>
  );
}

export default App;
