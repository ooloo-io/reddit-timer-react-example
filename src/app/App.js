import React from 'react';
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { Normalize } from 'styled-normalize';
import { GlobalStyle, theme } from '../style';
import Header from '../common/header';

function App() {
  return (
    <ThemeProvider theme={theme}>
      <Router>
        <Normalize />
        <GlobalStyle />
        <Header />
        <Switch>
          <Route path="/search">Search</Route>
          <Route path="/">Home</Route>
        </Switch>
      </Router>
    </ThemeProvider>
  );
}

export default App;
