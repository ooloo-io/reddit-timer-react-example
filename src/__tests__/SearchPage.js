import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';
import {
  render, screen, within, waitFor,
} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';
import App from '../app';
import { defaultSubreddit } from '../config';

const setup = (initialPath) => {
  // access history as described in the docs
  // https://reactrouter.com/web/guides/testing/checking-location-in-tests
  let history;
  const view = render(
    <MemoryRouter initialEntries={[initialPath]}>
      <App />
      <Route
        path="*"
        render={(props) => {
          history = props.history;
          return null;
        }}
      />
    </MemoryRouter>,
  );
  return { ...view, history };
};

async function clickFirstCellWithValue(value) {
  const heatmap = await screen.findByTestId('heatmap');
  const cell = within(heatmap).getAllByText(value)[0];
  userEvent.click(cell);
}

describe('subreddit form', () => {
  test('submitting the form updates the URL', async () => {
    const { history } = setup('/search/reactjs');

    const searchInput = screen.getByLabelText('r /');
    expect(searchInput.value).toBe('reactjs');

    userEvent.clear(searchInput);
    userEvent.type(searchInput, 'vuejs');
    expect(searchInput.value).toBe('vuejs');

    const submitButton = screen.getByRole('button', { name: 'Search' });
    userEvent.click(submitButton);

    expect(history.location.pathname).toEqual('/search/vuejs');

    await waitFor(() => expect(screen.queryByText('loading-spinner.svg')).not.toBeInTheDocument());
  });

  test('input value changes to default subreddit when search link in header is clicked', async () => {
    setup('/search/reactjs');
    const searchInput = screen.getByRole('textbox');
    const header = screen.getByRole('banner');
    const searchLink = within(header).getByRole('link', { name: /Search/ });

    userEvent.click(searchLink);

    expect(searchInput.value).toBe(defaultSubreddit);

    await waitFor(() => expect(screen.queryByText('loading-spinner.svg')).not.toBeInTheDocument());
  });
});

describe('heatmap', () => {
  test('loads and renders top posts for subreddit in URL', async () => {
    setup('/search/reactjs');

    screen.getByText('loading-spinner.svg');

    expect(await screen.findByTestId('heatmap')).toBeInTheDocument();
    expect(screen.queryByText('loading-spinner.svg')).not.toBeInTheDocument();

    const heatmap = screen.getByTestId('heatmap');
    const cells = await within(heatmap).findAllByRole('button');
    expect(cells.length).toEqual(7 * 24);

    const numberOfPostsPerCell = cells.map((cell) => cell.innerHTML);
    expect(numberOfPostsPerCell).toMatchSnapshot();

    const timezone = screen.getByText('All times are shown in your timezone:');
    expect(within(timezone).getByText('Europe/Berlin')).toBeInTheDocument();
  });

  test('cell highlights on click', async () => {
    setup('/search/reactjs');
    const heatmap = await screen.findByTestId('heatmap');
    const cells = await within(heatmap).findAllByRole('button');

    const cellToClick = cells[1];
    expect(cellToClick).toHaveStyle('border: none');

    userEvent.click(cellToClick);
    expect(cellToClick).toHaveStyle('border: 1px solid #1e2537');
  });

  test('renders error message', async () => {
    setup('/search/failing-request');

    expect(await screen.findByText(/something went wrong/i)).toBeInTheDocument();
    expect(screen.queryByText('loading-spinner.svg')).not.toBeInTheDocument();
  });
});

describe('posts table', () => {
  test('is not visible when no cell is clicked', async () => {
    setup('/search/reactjs');
    await screen.findByTestId('heatmap');

    expect(screen.queryByRole('table')).not.toBeInTheDocument();
  });

  test('is not visible when cell with no posts is clicked', async () => {
    setup('/search/reactjs');
    await clickFirstCellWithValue('0');

    expect(screen.queryByRole('table')).not.toBeInTheDocument();
  });

  test('shows posts ordered by time according to cell that is clicked', async () => {
    setup('/search/reactjs');
    await clickFirstCellWithValue('4');

    const table = screen.getByRole('table');
    const tableRows = within(table)
      .getAllByRole('row')
      .slice(1);

    const tableContent = tableRows.map((row) => {
      const cells = within(row).getAllByRole('cell');
      const titleLink = within(cells[0]).getByRole('link');
      const authorLink = within(cells[4]).getByRole('link');
      return {
        title: titleLink.innerHTML,
        href: titleLink.href,
        time: cells[1].innerHTML,
        score: cells[2].innerHTML,
        numComments: cells[3].innerHTML,
        author: authorLink.innerHTML,
        authorHref: authorLink.href,
      };
    });

    expect(tableContent).toMatchSnapshot();
  });

  test('shows no link for deleted user', async () => {
    setup('/search/reactjs');
    const heatmap = await screen.findByTestId('heatmap');
    const sunday5pm = within(heatmap).getAllByText('6')[1];
    userEvent.click(sunday5pm);

    const table = screen.getByRole('table');
    const rowWithDeletedUser = within(table).getAllByRole('row')[4];

    const authorCell = within(rowWithDeletedUser).getAllByRole('cell')[4];
    expect(within(authorCell).queryByRole('link')).not.toBeInTheDocument();
    expect(authorCell.innerHTML).toBe('[deleted]');
  });
});

test('no accessibility violations', async () => {
  const { container } = setup('/search/reactjs');
  await clickFirstCellWithValue('4');

  expect(await axe(container)).toHaveNoViolations();
});
