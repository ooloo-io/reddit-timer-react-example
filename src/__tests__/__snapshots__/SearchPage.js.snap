// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`heatmap loads and renders top posts for subreddit in URL 1`] = `
Array [
  "3",
  "1",
  "1",
  "0",
  "1",
  "1",
  "4",
  "0",
  "0",
  "3",
  "1",
  "1",
  "3",
  "1",
  "3",
  "3",
  "6",
  "6",
  "5",
  "5",
  "6",
  "1",
  "4",
  "2",
  "1",
  "1",
  "3",
  "1",
  "2",
  "4",
  "0",
  "3",
  "0",
  "3",
  "1",
  "3",
  "3",
  "7",
  "3",
  "6",
  "12",
  "3",
  "4",
  "7",
  "3",
  "4",
  "4",
  "1",
  "1",
  "4",
  "1",
  "0",
  "2",
  "0",
  "3",
  "1",
  "4",
  "6",
  "2",
  "3",
  "2",
  "3",
  "7",
  "8",
  "3",
  "8",
  "3",
  "5",
  "5",
  "3",
  "3",
  "1",
  "0",
  "2",
  "1",
  "1",
  "0",
  "3",
  "3",
  "2",
  "2",
  "1",
  "2",
  "1",
  "5",
  "5",
  "10",
  "9",
  "4",
  "4",
  "2",
  "5",
  "3",
  "6",
  "5",
  "4",
  "4",
  "1",
  "0",
  "0",
  "0",
  "0",
  "1",
  "2",
  "1",
  "0",
  "6",
  "1",
  "7",
  "2",
  "7",
  "7",
  "6",
  "8",
  "3",
  "6",
  "2",
  "0",
  "4",
  "4",
  "1",
  "5",
  "0",
  "2",
  "2",
  "0",
  "1",
  "4",
  "1",
  "0",
  "2",
  "1",
  "3",
  "2",
  "4",
  "7",
  "5",
  "5",
  "5",
  "1",
  "2",
  "6",
  "4",
  "1",
  "2",
  "2",
  "1",
  "2",
  "2",
  "1",
  "2",
  "0",
  "2",
  "0",
  "4",
  "2",
  "3",
  "2",
  "5",
  "5",
  "7",
  "4",
  "6",
  "2",
  "3",
  "3",
  "1",
  "5",
]
`;

exports[`posts table shows posts ordered by time according to cell that is clicked 1`] = `
Array [
  Object {
    "author": "mutantdustbunny",
    "authorHref": "https://reddit.com/u/mutantdustbunny",
    "href": "https://reddit.com/r/javascript/comments/fiwfrj/css_dark_mode_tutorial_adding_selectable_themes/",
    "numComments": "30",
    "score": "246",
    "time": "6:19 am",
    "title": "CSS Dark Mode tutorial â€“ adding selectable themes to your site or PWA with alternate style",
  },
  Object {
    "author": "hotpot_ai",
    "authorHref": "https://reddit.com/u/hotpot_ai",
    "href": "https://reddit.com/r/javascript/comments/fmuoc9/a_realworld_comparison_of_frontend_frameworks_2020/",
    "numComments": "94",
    "score": "211",
    "time": "6:24 am",
    "title": "A Real-World Comparison of Front-End Frameworks 2020",
  },
  Object {
    "author": "Quarxnox",
    "authorHref": "https://reddit.com/u/Quarxnox",
    "href": "https://reddit.com/r/javascript/comments/gckiuc/askjs_today_i_learned_of_the_textspeech_apis/",
    "numComments": "25",
    "score": "186",
    "time": "6:25 am",
    "title": "[AskJS] Today I learned of the text/speech apis.",
  },
  Object {
    "author": "unhandledrejection",
    "authorHref": "https://reddit.com/u/unhandledrejection",
    "href": "https://reddit.com/r/javascript/comments/hturg4/i_am_building_a_open_source_mindmap_editor/",
    "numComments": "25",
    "score": "151",
    "time": "6:36 am",
    "title": "I am building a open source mindmap editor",
  },
]
`;
